### PR TITLE
routing+gossiper: avoid unnecessarily validating sigs for announcement that is "stale"

### PR DESCRIPTION
**Pull request by Roasbeef**
In this PR, we reduce the amount of unnecessary work that the gossiper can carry out. When CPU profiling some nodes, I noticed that we’d spend a lot of time validating the signatures for an announcement, only to realize that the router already had it.

To remedy this, we’ll use the new methods added to the channel router in order to avoid unnecessarily validating an announcement that is actually stale. This should reduce memory usage (since it uses big int’s under the scenes), and also idle CPU usage.

The new methods added to the channel router's `ChannelGraphSource` interface are: 
```go
 
// IsStaleNode returns true if the graph source has a node announcement
// for the target node with a more recent timestamp.
IsStaleNode(node Vertex, timestamp time.Time) bool
 
// IsKnownEdge returns true if the graph source already knows of the
// passed channel ID.
IsKnownEdge(chanID lnwire.ShortChannelID) bool
 
// IsStaleEdgePolicy returns true if the graph source has a channel
// edge for the passed channel ID (and flags) that have a more recent
// timestamp.
IsStaleEdgePolicy(chanID lnwire.ShortChannelID, timestamp time.Time,
flags lnwire.ChanUpdateFlag) bool
```

### CODE REVIEW

File: routing/router.go [hunk 1]
```diff
@@ -1907,3 +1937,65 @@ func (r *ChannelRouter) AddProof(chanID lnwire.ShortChannelID,
...
+	// A flag set of 0 indicates this is an announcement for the "first"
+	// node in the channel.
+	case flags&lnwire.ChanUpdateDirection == 0:
+		return edge1Timestamp.After(timestamp) ||
+			edge1Timestamp.Equal(timestamp)
+
+	// Similarly, a flag set of 1 indicates this is an announcement for the
+	// "second" node in the channel.
+	case flags&lnwire.ChanUpdateDirection == 1:
+		return edge2Timestamp.After(timestamp) ||
```
Comment (nalinbhardwaj) [inline routing/router.go]: nit: this is more clear to me as ->
```
return !edge2Timestamp.Before(timestamp)
```
Same for other uses of timestamps

Comment (Roasbeef) [reply routing/router.go]: Agreed, will update!

Comment (nalinbhardwaj) [reply routing/router.go]: I think you forgot to modify the test files(example `gossiper_test.go`).

File: routing/router.go [hunk 2]
```diff
@@ -819,6 +833,42 @@ func (r *ChannelRouter) networkHandler() {
...
+	// node announcements, we will ignore such nodes. If we do know about
+	// this node, check that this update brings info newer than what we
+	// already have.
+	lastUpdate, exists, err := r.cfg.Graph.HasLightningNode(node)
+	if err != nil {
+		return errors.Errorf("unable to query for the "+
+			"existence of node: %v", err)
+	}
+	if !exists {
+		return newErrf(ErrIgnored, "Ignoring node announcement"+
```
Comment (halseth) [inline routing/router.go]: The godoc doesn't mention this error.

File: routing/router.go [hunk 3]
```diff
@@ -1907,3 +1937,65 @@ func (r *ChannelRouter) AddProof(chanID lnwire.ShortChannelID,
...
 }
+
+// IsStaleNode returns true if the graph source has a node announcement for the
+// target node with a more recent timestamp.
+//
+// NOTE: This method is part of the ChannelGraphSource interface.
+func (r *ChannelRouter) IsStaleNode(node Vertex, timestamp time.Time) bool {
+	// If our attempt to assert that the node announcement is fresh fails,
+	// then we know that this is actually a stale announcement.
+	return r.assertNodeAnnFreshness(node, timestamp) != nil
```
Comment (halseth) [inline routing/router.go]: If the node does not exist, then this will fail?

File: routing/router_test.go [hunk 4]
```diff
@@ -1387,3 +1387,228 @@ func TestFindPathFeeWeighting(t *testing.T) {
...
+		HaveNodeAnnouncement: true,
+		LastUpdate:           updateTimeStamp,
+		Addresses:            testAddrs,
+		Color:                color.RGBA{1, 2, 3, 0},
+		Alias:                "node11",
+		AuthSigBytes:         testSig.Serialize(),
+		Features:             testFeatures,
+	}
+	copy(n1.PubKeyBytes[:], priv1.PubKey().SerializeCompressed())
+	if err := ctx.router.AddNode(n1); err != nil {
```
Comment (halseth) [inline routing/router_test.go]: Should also test the `IsStaleNode` method before adding the node.

File: routing/router_test.go [hunk 5]
```diff
@@ -1387,3 +1387,228 @@ func TestFindPathFeeWeighting(t *testing.T) {
...
+	// If we update the timestamp and once again query for staleness, it
+	// should report false.
+	newTimeStamp := time.Unix(1234, 0)
+	if ctx.router.IsStaleNode(pub1, newTimeStamp) {
+		t.Fatalf("incorrectly detected node as stale")
+	}
+}
+
+// TestIsKnownEdge tests that the IsKnownEdge method properly detects stale
+// node announcements.
```
Comment (halseth) [inline routing/router_test.go]: channel announcements

File: routing/router_test.go [hunk 6]
```diff
@@ -1387,3 +1387,228 @@ func TestFindPathFeeWeighting(t *testing.T) {
...
+
+	edge := &channeldb.ChannelEdgeInfo{
+		ChannelID:        chanID.ToUint64(),
+		NodeKey1Bytes:    pub1,
+		NodeKey2Bytes:    pub2,
+		BitcoinKey1Bytes: pub1,
+		BitcoinKey2Bytes: pub2,
+		AuthProof:        nil,
+	}
+	if err := ctx.router.AddEdge(edge); err != nil {
```
Comment (halseth) [inline routing/router_test.go]: should also here test the method before adding the edge.

File: routing/router.go (hunk 1 above)
Comment (Roasbeef) [reply routing/router.go]: Fixed!

File: routing/router.go (hunk 2 above)
Comment (Roasbeef) [reply routing/router.go]: Fixed.

File: routing/router.go (hunk 3 above)
Comment (Roasbeef) [reply routing/router.go]: Yep, as we need the _partial_ node announcement to be in the channel graph (added after we get a chan ann) before we'll accept the full node announcement. We'll insert a partial node when we get the channel announcement, and the full node (addrs, etc) once we get the chan ann.

File: routing/router_test.go (hunk 4 above)
Comment (Roasbeef) [reply routing/router_test.go]: Done.

File: routing/router_test.go (hunk 5 above)
Comment (Roasbeef) [reply routing/router_test.go]: Fixed.

File: routing/router_test.go (hunk 6 above)
Comment (Roasbeef) [reply routing/router_test.go]: Fixed.

File: routing/router.go [hunk 7]
```diff
@@ -819,6 +833,42 @@ func (r *ChannelRouter) networkHandler() {
...
+		return newErrf(ErrIgnored, "Ignoring node announcement"+
+			" for node not found in channel graph (%x)",
+			node[:])
+	}
+
+	// If we've reached this point then we're aware of the vertex being
+	// advertised. So we now check if the new message has a new time stamp,
+	// if not then we won't accept the new data as it would override newer
+	// data.
+	if exists && lastUpdate.After(msgTimestamp) ||
```
Comment (halseth) [inline routing/router.go]: `exists` is redundant here

File: routing/router.go (hunk 3 above)
Comment (halseth) [reply routing/router.go]: This is counterintuitive imo, as the godoc states that the method returns true if we have already seen a newer node announcement, but the real behaviour is that it returns true if `(we have seen newer node announcement || we have never seen a channel announcement from this node)`.

Either the godoc and method names should be updated, or the behaviour should be changed to reflect the doc and usage be 

```go
onReceiveNodeAnnouncement(ann) {
    if !existsInGraph(ann) {
        return ignoreNode
    }

    if IsStaleNode(ann) {
        return ignoreNode
    }

    apply(ann)
}
```

**Comment by Roasbeef** [comment]
@halseth updated the `IsStaleNode` `godoc`.

File: routing/router.go (hunk 3 above)
Comment (Roasbeef) [reply routing/router.go]: Fail as in return true? Yes

**Pull request merged**